include/cuco/detail/static_set/kernels.cuh

```diff
@@ -89,6 +90,257 @@ CUCO_KERNEL void find(InputIt first, cuco::detail::index_type n, OutputIt output
   }
 }
 
+template <typename CG,
+          typename Size,  // have to do this due to https://github.com/NVIDIA/cccl/issues/1523
+          typename ProbeKey,
+          typename Key,
+          typename AtomicT,
+          typename OutputIt1,
+          typename OutputIt2>
+__device__ void flush_buffer(CG const& tile,
+                             Size buffer_size,
+                             cuco::pair<ProbeKey, Key>* buffer,
+                             AtomicT* counter,
+                             OutputIt1 output_probe,
+                             OutputIt2 output_match)
+{
+  Size offset;
+  auto i = tile.thread_rank();
+
+#if defined(CUCO_HAS_CG_INVOKE_ONE)
+  offset = cooperative_groups::invoke_one_broadcast(
+    tile, [&]() { return counter->fetch_add(buffer_size, cuda::std::memory_order_relaxed); });
+#else
+  if (i == 0) { offset = counter->fetch_add(buffer_size, cuda::std::memory_order_relaxed); }
+  offset = tile.shfl(offset, 0);
+#endif
+
+  while (i < buffer_size) {
+    *(output_probe + offset + i) = buffer[i].first;
+    *(output_match + offset + i) = buffer[i].second;
+
+    i += tile.size();
+  }
+}
+
+template <int32_t BlockSize,
+          typename InputIt,
+          typename OutputIt1,
+          typename OutputIt2,
+          typename AtomicT,
+          typename Ref>
+__device__ void group_retrieve(InputIt first,
+                               cuco::detail::index_type n,
+                               OutputIt1 output_probe,
+                               OutputIt2 output_match,
+                               AtomicT* counter,
+                               Ref ref)
+{
+  namespace cg = cooperative_groups;
+
+  using size_type = typename Ref::size_type;
+  using ProbeKey  = typename std::iterator_traits<InputIt>::value_type;
+  using Key       = typename Ref::key_type;
+
+  auto constexpr tile_size   = Ref::cg_size;
+  auto constexpr window_size = Ref::window_size;
+
+  auto idx          = cuco::detail::global_thread_id() / tile_size;
+  auto const stride = cuco::detail::grid_stride() / tile_size;
+  auto const block  = cg::this_thread_block();
+  auto const tile   = cg::tiled_partition<tile_size>(block);
+
+  auto constexpr flushing_tile_size = cuco::detail::warp_size() / window_size;
+  // random choice to tune
+  auto constexpr flushing_buffer_size = 2 * flushing_tile_size;
```

Comment:
I'm curious. Why did you choose that particular size?

Reply:
No particular reason. Tested with 1, 2, 3 and 4 and there is no big difference between those options.